### PR TITLE
Fix revoke blob url

### LINKED ISSUE #24290: URL.RevokeBlobURL is not strict enough about blob urls
Spec: https://w3c.github.io/FileAPI/#dfn-revokeObjectURL

Code:
* `components/script/dom/url.rs`
* `components/net_traits/blob_url_store.rs`

Test:
* `./mach test-wpt tests/wpt/web-platform-tests/FileAPI/url/url-with-xhr.any.js`
* `./mach test-wpt tests/wpt/web-platform-tests/FileAPI/url/url-with-fetch.any.js`

There are a couple problems:
* we don't do any same-origin check against the blob being revoked
* the `parse_blob_url` helper strips off any extra parts of the URL being revoked, whereas https://w3c.github.io/FileAPI/#removeTheEntry serializes the entire url

### PR DESCRIPTION
**Pull request by shnmorimoto**
fix #24290

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #24290 (GitHub issue number if applicable)

- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because ___

### CODE REVIEW

**Comment by jdm** [comment]
Hmm, the goal of #24290 was to get tests/wpt/web-platform-tests/FileAPI/url/url-with-xhr.any.js and tests/wpt/web-platform-tests/FileAPI/url/url-with-fetch.any.js passing, so it's curious that there is no change in test results.

File: components/net/filemanager_thread.rs [hunk 1]
```diff
@@ -163,8 +163,10 @@ impl FileManager {
             FileManagerThreadMsg::DecRef(id, origin, sender) => {
                 let _ = sender.send(self.store.dec_ref(&id, &origin));
             },
-            FileManagerThreadMsg::RevokeBlobURL(id, origin, sender) => {
-                let _ = sender.send(self.store.set_blob_url_validity(false, &id, &origin));
+            FileManagerThreadMsg::RevokeBlobURL(url, sender) => {
+                if let Ok((id, origin)) = parse_blob_url(&url) {
```
Comment (jdm) [inline components/net/filemanager_thread.rs]: Oh - we still are throwing away the fragment and query string from the URL, which is why the tests continue to fail. We need to explicitly deal with those.

Comment (shnmorimoto) [reply components/net/filemanager_thread.rs]: Thanks for the review.
I will check and fix it!

**Comment by shnmorimoto** [comment]
@jdm 

I have two quesrtions.

### 1. About parse_blob_url

parse_blob_url gets uuid and origin from url.

ex.
```
if "blob:http://web-platform.test:8000/8da840d9fed84875be1e896b8b2cbcc9" is given as url.
And parse_blob_url gets 8da840d9fed84875be1e896b8b2cbcc9 as uuid and http://web-platform.test:8000 as origin.
```

And, CreateObjectURL seems to store uuid as key.(because it call set_blob_url_validity.)
https://github.com/servo/servo/blob/e5689df6b44696bf24f8757cd1f6e14d1aabf6f0/components/net/filemanager_thread.rs#L679
        
So I think that it is correct that RevokeObjectURL uses uuid as key and uses parse_blob_url.

### 2. About failed tests.

I ran the test in my local(macOS Mojave 10.14.6).
And, I got the following results.

```
> ./mach test-wpt tests/wpt/web-platform-tests/FileAPI/url/url-with-xhr.any.js    

...

web-platform-test
~~~~~~~~~~~~~~~~~
Ran 30 checks (28 subtests, 2 tests)
Expected results: 30
Unexpected results: 0
OK
 0:05.78 INFO Closing logging queue
 0:05.78 INFO queue closed
```

```
> ./mach test-wpt tests/wpt/web-platform-tests/FileAPI/url/url-with-fetch.any.js

...

0:04.77 TEST_END: Test OK. Subtests passed 10/15. Unexpected 1
FAIL Revoke blob URL after calling fetch, fetch should succeed - promise_test: Unhandled rejection with value: object "TypeError: Network error occurred"
 0:04.77 INFO No more tests
 0:04.78 INFO Closing logging queue
 0:04.78 INFO queue closed
 0:04.80 TEST_END: Test OK. Subtests passed 10/15. Unexpected 1
FAIL Revoke blob URL after calling fetch, fetch should succeed - promise_test: Unhandled rejection with value: object "TypeError: Network error occurred"
 0:04.80 INFO No more tests
 0:04.81 INFO Closing logging queue
 0:04.81 INFO queue closed
 0:04.82 INFO Got 2 unexpected results
 0:04.82 SUITE_END

web-platform-test
~~~~~~~~~~~~~~~~~
Ran 32 checks (30 subtests, 2 tests)
Expected results: 30
Unexpected results: 2
  subtest: 2 (2 fail)

Unexpected Results
------------------
/FileAPI/url/url-with-fetch.any.html
  FAIL Revoke blob URL after calling fetch, fetch should succeed - promise_test: Unhandled rejection with value: object "TypeError: Network error occurred"
/FileAPI/url/url-with-fetch.any.worker.html
  FAIL Revoke blob URL after calling fetch, fetch should succeed - promise_test: Unhandled rejection with value: object "TypeError: Network error occurred"
```

Is this the error you said?

And, I tried adding the following to url-with-fetch.any.js.

```javascript

  ...
  const result = fetch_should_succeed(t, url).then(text => {
    assert_equals(text, blob_contents);
  });

  // added code
  for (i = 0; i < 1000; i ++) {
    console.log("aaaa");
  }

  // Revoke the object URL. fetch should have already resolved the blob URL.
  URL.revokeObjectURL(url);

  return result;
}, 'Revoke blob URL after calling fetch, fetch should succeed');

```

And, I got the following result.

```
> ./mach test-wpt tests/wpt/web-platform-tests/FileAPI/url/url-with-fetch.any.js
...

 0:05.20 TEST_END: Test OK. Subtests passed 11/15. Unexpected 0
 0:05.20 INFO No more tests
 0:05.21 INFO Closing logging queue
 0:05.21 INFO queue closed
 0:05.21 INFO Got 0 unexpected results
 0:05.21 SUITE_END

web-platform-test
~~~~~~~~~~~~~~~~~
Ran 32 checks (30 subtests, 2 tests)
Expected results: 32
Unexpected results: 0
OK
 0:05.24 INFO Closing logging queue
 0:05.24 INFO queue closed
```

So, I think that the cause of the error is the fetch timing.

Is this correct?

**Comment by jdm** [comment]
@shnmorimoto With respect to the test output, when a test says:
```
Ran 30 checks (28 subtests, 2 tests)
Expected results: 30
Unexpected results: 0
OK
 0:05.78 INFO Closing logging queue
 0:05.78 INFO queue closed
```
That means that the test results matched the expected results. For that test, they are present in [this file](https://github.com/servo/servo/blob/master/tests/wpt/metadata/FileAPI/url/url-with-xhr.any.js.ini), which shows that various subtests are currently expected to fail. By making the changes I recommend, our goal is to make the subtests pass and remove these expected failures.

As for the question about parse_blob_url, rather than my original suggestion (make the RevokeBlobURL message accept a URL argument), I think we should modify parse_blob_url to return an Err if the parsed URL contains a fragment or query string. This should achieve the same goal - revoking a blob URL that has a fragment (`#foo`) attached to it should not cause the blob to be revoked.

**Comment by shnmorimoto** [comment]
@jdm I misunderstood the test output. Thank you for the detailed explanation.

> I think we should modify parse_blob_url to return an Err if the parsed URL contains a fragment or query string. This should achieve the same goal - revoking a blob URL that has a fragment (#foo) attached to it should not cause the blob to be revoked.

I understand. I will fix it that way.

**Comment by shnmorimoto** [comment]
@jdm I've fixed it. For fragment, there are following [test](https://github.com/servo/servo/blob/a7008d1cf71c6bcdf49ce41d45178a413199c837/tests/wpt/web-platform-tests/FileAPI/url/resources/fetch-tests.js#L28).

```javascript
 promise_test(t => {
    const url = URL.createObjectURL(blob);

    return fetch_should_succeed(t, url + '#fragment').then(text => {
      assert_equals(text, blob_contents);
    });
  }, fetch_method + ' with a fragment should succeed');
```

So, It is only checked in RevokeObjectURL.

What do you think?

**Comment by jdm** [comment]
Good find. I agree!

**Comment by shnmorimoto** [comment]
Thanks! I fixed it as follows and already pushed.

```rust
if let Ok(url) = ServoUrl::parse(&url) {
    if url.fragment().is_none() && origin == get_blob_origin(&url) {
```
I ran the test locally and confirmed that the results changed.

```
web-platform-test
~~~~~~~~~~~~~~~~~
Ran 30 checks (28 subtests, 2 tests)
Expected results: 24
Unexpected results: 6
  subtest: 6 (6 pass)

Unexpected Results
------------------
/FileAPI/url/url-with-xhr.any.worker.html
  UNEXPECTED-PASS Only exact matches should revoke URLs, using XHR
  UNEXPECTED-PASS Appending a query string should cause XHR to fail
  UNEXPECTED-PASS Appending a path should cause XHR to fail
/FileAPI/url/url-with-xhr.any.html
  UNEXPECTED-PASS Only exact matches should revoke URLs, using XHR
  UNEXPECTED-PASS Appending a query string should cause XHR to fail
  UNEXPECTED-PASS Appending a path should cause XHR to fail
 0:05.07 INFO Closing logging queue
 0:05.08 INFO queue closed
```

**Comment by jdm** [comment]
Please include the [updated test results](https://github.com/servo/servo/blob/master/tests/wpt/README.md#updating-test-expectations) in this PR, in that case :)

**Comment by jdm** [comment]
You can use `./mach test-wpt tests/wpt/web-platform-tests/FileAPI` as the base command in order to avoid having to run thousands of unnecessary tests, too!

**Comment by shnmorimoto** [comment]
Oh, Sorry. I updated test results.

I still get unexpected results locally.

```
> ./mach test-wpt tests/wpt/web-platform-tests/FileAPI                                                                                                                  fix_revoke_blob_url_2 ✱
Running 56 tests in web-platform-tests

  ▶ Unexpected subtest result in /FileAPI/url/url-with-fetch.any.worker.html:
  │ FAIL [expected PASS] Revoke blob URL after calling fetch, fetch should succeed
  └   → promise_test: Unhandled rejection with value: object "TypeError: Network error occurred"

  ▶ Unexpected subtest result in /FileAPI/url/url-in-tags-revoke.window.html:
  │ FAIL [expected TIMEOUT] Opening a blob URL in a new window immediately before revoking it works.
  │   → assert_equals: expected (string) "test_frame_OK" but got (undefined) undefined
  │
  │ win.onload<@http://web-platform.test:8000/FileAPI/url/url-in-tags-revoke.window.js:52:5
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1908:25
  └ Test.prototype.step_func_done/<@http://web-platform.test:8000/resources/testharness.js:1948:32

  ▶ Unexpected subtest result in /FileAPI/url/url-with-fetch.any.html:
  │ FAIL [expected PASS] Revoke blob URL after calling fetch, fetch should succeed
  └   → promise_test: Unhandled rejection with value: object "TypeError: Network error occurred"

Ran 56 tests finished in 117.0 seconds.
  • 53 ran as expected. 0 tests skipped.
  • 3 tests had unexpected subtest results  
```

But, I get same results with master branch too.
So, I ignored it this time.

**Comment by shnmorimoto** [comment]
I did rebase and resolved merge conflicts.

**Comment by jdm** [comment]
@bors-servo r+
Thanks for fixing this!

**Comment by jdm** [comment]
@bors-servo retry
* #24762

**Pull request merged**